src/k2/module/k23si/TxnManager.cpp

```diff
@@ -192,247 +199,479 @@ TxnRecord& TxnManager::_createRecord(dto::TxnId txnId) {
 
         _hblist.push_back(rec);
         _rwlist.push_back(rec);
+        K2LOG_D(log::skvsvr, "created new txn record: {}", it.first->second);
+    }
+    else {
+        K2LOG_D(log::skvsvr, "found existing txn record: {}", it.first->second);
     }
-    K2LOG_D(log::skvsvr, "created new txn record: {}", it.first->second);
     return it.first->second;
 }
 
-seastar::future<> TxnManager::onAction(TxnRecord::Action action, dto::TxnId txnId) {
-    // This method's responsibility is to execute valid state transitions.
-    TxnRecord& rec = getTxnRecord(std::move(txnId));
+seastar::future<Status> TxnManager::createTxn(dto::TxnId&& txnId) {
+    return _onAction(TxnRecord::Action::onCreate, getTxnRecord(std::move(txnId)));
+}
+
+seastar::future<Status> TxnManager::heartbeat(dto::TxnId&& txnId) {
+    return _onAction(TxnRecord::Action::onHeartbeat, getTxnRecord(std::move(txnId)));
+}
+
+// return true if the challenge was successful (challenger wins over incumbent)
+bool _evaluateChallenge(TxnRecord& incumbent, dto::K23SI_MTR& challengerMTR) {
+    // Calculate if the incumbent would lose the challenge based on conflict resolution
+    bool incumbentLostConflict = false;
+    // #1 abort based on priority
+    if (incumbent.txnId.mtr.priority > challengerMTR.priority) {  // bigger number means lower priority
+        K2LOG_D(log::skvsvr, "incumbent {} could lose push", incumbent.txnId);
+        incumbentLostConflict = true;
+    }
+    // #2 if equal, pick the newer transaction
+    else if (incumbent.txnId.mtr.priority == challengerMTR.priority) {
+        auto cmpResult = incumbent.txnId.mtr.timestamp.compareCertain(challengerMTR.timestamp);
+        if (cmpResult == dto::Timestamp::LT) {
+            K2LOG_D(log::skvsvr, "incumbent {} could lose push", incumbent.txnId);
+            incumbentLostConflict = true;
+        } else if (cmpResult == dto::Timestamp::EQ) {
+            // #3 if same priority and timestamp, abort on tso ID which must be unique
+            if (incumbent.txnId.mtr.timestamp.tsoId() < challengerMTR.timestamp.tsoId()) {
```

Comment:
compareCertain already considers the tsoID, so I think you can remove these branches.